server/asset/eth/coiner.go

```diff
@@ -57,35 +47,77 @@ type swapCoin struct {
-// an error is returned then if something is different than expected. As such,
-// the swapCoin expects Confirmations to be called with confirmations
-// available at least once before the swap be trusted for swap initializations.
-func (backend *Backend) newSwapCoin(coinID []byte, contractData []byte, sct swapCoinType) (*swapCoin, error) {
-	switch sct {
-	case sctInit, sctRedeem:
-	default:
-		return nil, fmt.Errorf("unknown swapCoin type: %d", sct)
+// fields are verified when the Confirmations method is called.
+func (eth *Backend) newSwapCoin(coinID []byte, contractData []byte) (*swapCoin, error) {
+	bc, err := eth.baseCoin(coinID, contractData)
+	if err != nil {
+		return nil, err
 	}
 
+	inits, err := dexeth.ParseInitiateData(bc.txData, ethContractVersion)
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse initiate call data: %v", err)
+	}
+
+	init, ok := inits[bc.secretHash]
+	if !ok {
+		return nil, fmt.Errorf("tx %v does not contain initiation with secret hash %x", bc.txHash, bc.secretHash)
+	}
+
+	var sum uint64
+	for _, in := range inits {
+		sum += in.Value
+	}
+	if bc.value < sum {
+		return nil, fmt.Errorf("tx %s value < sum of inits. %d < %d", bc.txHash, bc.value, sum)
+	}
+
+	return &swapCoin{
+		baseCoin: bc,
+		init:     init,
+	}, nil
+}
+
+// newRedeemCoin pulls the tx for the coinID => txHash, extracts the secret, and
+// provides a coin to check confirmations, as required by asset.Coin interface,
+// TODO: The redeemCoin's Confirmation method is never used by the current
+// swapper implementation. Might consider an API change for
+// asset.Backend.Redemption.
+func (eth *Backend) newRedeemCoin(coinID []byte, contractData []byte) (*redeemCoin, error) {
+	bc, err := eth.baseCoin(coinID, contractData)
+	if err != nil {
+		return nil, err
+	}
+
+	redemptions, err := dexeth.ParseRedeemData(bc.txData, ethContractVersion)
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse redemption call data: %v", err)
+	}
+	redemption, ok := redemptions[bc.secretHash]
+	if !ok {
+		return nil, fmt.Errorf("tx %v does not contain redemption with secret hash %x", bc.txHash, bc.secretHash)
+	}
+
+	return &redeemCoin{
+		baseCoin: bc,
+		secret:   redemption.Secret,
+	}, nil
+}
+
+// The baseCoin is basic tx and swap contract data.
+func (eth *Backend) baseCoin(coinID []byte, contractData []byte) (*baseCoin, error) {
 	txHash, err := dexeth.DecodeCoinID(coinID)
 	if err != nil {
 		return nil, err
 	}
-	tx, _, err := backend.node.transaction(backend.rpcCtx, txHash)
+	tx, _, err := eth.node.transaction(eth.rpcCtx, txHash)
 	if err != nil {
 		if errors.Is(err, ethereum.NotFound) {
 			return nil, asset.CoinNotFoundError
 		}
 		return nil, fmt.Errorf("unable to fetch transaction: %v", err)
 	}
 
-	txdata := tx.Data()
-	// Transactions that call contract functions must have extra data.
-	if len(txdata) == 0 {
-		return nil, errors.New("tx calling contract function has no extra data")
-	}
-
```

Comment:
Was this check not accurate? I thought it was a good way to fail fast.

Reply:
`baseCoin` is more generalized. The txdata length check is now performed (via `ParseInitiateData`) in `newSwapCoin` immediately after the `baseCoin` is created.